### PR TITLE
Pbs cluster

### PR DESCRIPTION
**Pull request by AdamWlodarczyk**
PBS custom cluster proof of concept. It is a base for more wide usage.

### CODE REVIEW

File: jobqueue_features/clusters.py [hunk 1]
```diff
@@ -687,4 +703,41 @@ def _update_script_nodes(self, **kwargs) -> None:
...
+                    "We don't allow 'job_cpu' as a kwarg in MPI mode since we"
+                    " leverage this in jobqueue to set cpus-per-task"
+                )
+            mpi_job_extra = []
+            if "job_extra" not in kwargs:
+                raise KeyError("job_extra keyword should always be set in kwargs")
+            mpi_job_extra.extend(kwargs["job_extra"])
+            kwargs.update({"job_extra": mpi_job_extra})
+        super().__init__(**kwargs)
+        self._kwargs["mpi_tasks"] = self.mpi_tasks
```
Comment (ocaisa) [inline jobqueue_features/clusters.py]: This only exists in `mpi_mode` so you need a protected assignment

**Comment by ocaisa** [comment]
FYI I tried the example out in CI and it works...except you don't get the PBS output files back (because we haven't configured a shared home, and for us it is not worth the effort).

File: jobqueue_features/clusters.py [hunk 2]
```diff
@@ -562,15 +611,21 @@ def _update_kwargs_env_extra(self, **kwargs) -> Dict[str, Any]:
...
-    """Custom SLURMCluster class with CustomClusterMixin for initial kwargs tweak.
-     adds client attribute to SLURMCluster class."""
+    __doc__ = f"""Custom SLURMCluster class with CustomClusterMixin for initial kwargs tweak.
+
+     Adds client attribute to SLURMCluster class.
+
+     Attributes
+     ----------
+     {custom_cluster_attributes}
+     """
```
Comment (ocaisa) [inline jobqueue_features/clusters.py:696]: I'm probably showing my ignorance here but shouldn't this be
```
__doc__ = f"""...""".format{custom_cluster_attributes=custom_cluster_attributes}
```

File: jobqueue_features/clusters.py [hunk 3]
```diff
@@ -687,4 +727,54 @@ def _update_script_nodes(self, **kwargs) -> None:
...
+                    "We don't allow 'job_cpu' as a kwarg in MPI mode since we"
+                    " leverage this in jobqueue to set cpus-per-task"
+                )
+            mpi_job_extra = []
+            if "job_extra" not in kwargs:
+                raise KeyError("job_extra keyword should always be set in kwargs")
+            mpi_job_extra.extend(kwargs["job_extra"])
+            kwargs.update({"job_extra": mpi_job_extra})
+        super().__init__(**kwargs)
+        if hasattr(self, "mpi_tasks"):
```
Comment (ocaisa) [inline jobqueue_features/clusters.py]: This should probably be protected since it is only relevant in `mpi_mode`

**Review by ocaisa — CHANGES_REQUESTED** [review]
@AdamWlodarczyk Few minor comments but this looks good to me. The real test will be when we add the `mpi_mode` tests.

**Comment by ocaisa** [comment]
I ran the tests in #60 with this and I see where the problem is, in non-forked MPI mode you end up with a jobscript like:
```
#!/usr/bin/env bash

#PBS -N mpiMultiCluster
#PBS -q workq
#PBS -l select=1:ncpus=2:mpiprocs=2
#PBS -l walltime=00:01:00
export OMPI_ALLOW_RUN_AS_ROOT_CONFIRM=1
export OMPI_ALLOW_RUN_AS_ROOT=1
/opt/anaconda/bin/python -m distributed.cli.dask_worker tcp://172.19.0.2:35339 --nthreads 1 --memory-limit 256.00MB --name name --no-nanny --death-timeout 60 --local-directory /tmp
```

Notice the last line hasn't been modified to run with MPI.

File: jobqueue_features/clusters.py (hunk 2 above)
Comment (AdamWlodarczyk) [reply jobqueue_features/clusters.py:696]: This is an f-string so the `.format(foo=bar)` or `.format(bar)` is unnecessary.

File: jobqueue_features/clusters.py (hunk 3 above)
Comment (AdamWlodarczyk) [reply jobqueue_features/clusters.py]: Good point. I will add that.

**Comment by AdamWlodarczyk** [comment]
> I ran the tests in #60 with this and I see where the problem is, in non-forked MPI mode you end up with a jobscript like:
> 
> ```
> #!/usr/bin/env bash
> 
> #PBS -N mpiMultiCluster
> #PBS -q workq
> #PBS -l select=1:ncpus=2:mpiprocs=2
> #PBS -l walltime=00:01:00
> export OMPI_ALLOW_RUN_AS_ROOT_CONFIRM=1
> export OMPI_ALLOW_RUN_AS_ROOT=1
> /opt/anaconda/bin/python -m distributed.cli.dask_worker tcp://172.19.0.2:35339 --nthreads 1 --memory-limit 256.00MB --name name --no-nanny --death-timeout 60 --local-directory /tmp
> ```
> 
> Notice the last line hasn't been modified to run with MPI.

Thanks, I will look into it.

**Comment by ocaisa** [comment]
Ok, I got it to work, I'll make a PR to your branch.

**Comment by ocaisa** [comment]
Fix in #63 , makes tests in #60 pass

File: jobqueue_features/clusters.py (hunk 3 above)
Comment (AdamWlodarczyk) [reply jobqueue_features/clusters.py]: Changed in https://github.com/E-CAM/jobqueue_features/pull/52/commits/337d29f8c02c9eeeb926c293f1b2fcc22f822105

File: jobqueue_features/clusters.py [hunk 4]
```diff
@@ -614,77 +744,57 @@ def __del__(self):
...
+            if "job_extra" not in kwargs:
+                raise KeyError("job_extra keyword should always be set in kwargs")
+            mpi_job_extra.extend(kwargs["job_extra"])
+            kwargs.update({"job_extra": mpi_job_extra})
+        super().__init__(**kwargs)
+        self._update_script_nodes(**kwargs)
+        if self.mpi_mode:
+            if hasattr(self, "mpi_tasks"):
+                self._kwargs["mpi_tasks"] = self.mpi_tasks
+            if self.ngpus_per_node > 0:
```
Comment (ocaisa) [inline jobqueue_features/clusters.py]: Actually, the GPUs you might want to request outside of MPI mode.

**Review by ocaisa — CHANGES_REQUESTED** [review]
One last thing and then good to go

File: jobqueue_features/clusters.py (hunk 4 above)
Comment (AdamWlodarczyk) [reply jobqueue_features/clusters.py]: Changed it.

**Review by ocaisa — APPROVED** [review]
Just tested this with #60 and passses there as well, so good to go

**Comment by ocaisa** [comment]
Thanks @AdamWlodarczyk !

**Pull request merged**